### DISCUSSION TITLE
Process Volume and Pan Position commands

### DISCUSSION THREAD

**Discussion by Coozje** [cifkao/html-midi-player#61] [Ideas]
Hi,

I use your player as part of a midi editor I am building. The project is about halfway finished, and I only now discovered that the player doesn't process volume and pan position commands (Bn 07 xx and Bn 0A xx respectively). All notes sound equally loud and the music is mono.
This is a disappointment - there's little use for me in continuing if this is not fixed. Is there any chance the player will be updated anytime soon so that it can handle volume and pan position commands? That would really be so much appreciated. Thanks in advance!

**Comment by cifkao** [comment]
Unfortunately, this is one of the things that can only be fixed in [@magenta/music](https://github.com/magenta/magenta-js). You can try raising a request there. However, volume and pan are not supported by the `NoteSequence` protobuffer (see the definition of `Note` [here](https://github.com/magenta/note-seq/blob/55e4432a6686cec84b392c2290d4c2a1d040675c/note_seq/protobuf/music.proto#L104)), so implementing this would mean extending the protobuffer definition.

**Comment by danjferg** [comment]
My workaround/hack for volume:
- Add a listener to midi-player for the load event.
- On load, loop the note sequence notes and ...
    - find the max velocity. Save it as a default value.
    - create a list of all program and isDrum combinations.
    - for each combination, load all possible samples into the buffer using `player.player.loadAllSamples(program, isDrum);`
- Add a slider for Volume and adjust the velocities of all notes using a multiplication scale that maxes out when max velocity reaches 100.

Not a pure or perfect method, but a hack that met my needs for most cases.
